### PR TITLE
Solidity unpack

### PR DESCRIPTION
**Pull request by mimiiiz**
**SolidityUnpackString**
Decode data string from solidity encoded event data into given types.
Currently this function supports uint8, uint32, uint256, address, string data types.

### CODE REVIEW

File: common/evmcompat/helpers_test.go [hunk 1]
```diff
@@ -318,3 +321,117 @@ func TestAnotherSoliditySha3WithUnit64(t *testing.T) {
...
+func TestSolidityUnpackString(t *testing.T) {
+
+	types := []string{"uint8", "uint32", "address", "string", "string", "string", "uint256"}
+	data := "0x00000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000003000000000000000000000000989b462ebdda5e6ee4e05dd2884932ef44f920c300000000000000000000000000000000000000000000000000000000000000e00000000000000000000000000000000000000000000000000000000000000120000000000000000000000000000000000000000000000000000000000000016000000000000000000000000000000000000000000000000010f15cb27f673d5c0000000000000000000000000000000000000000000000000000000000000004616161610000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000046d696d690000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000046465657000000000000000000000000000000000000000000000000000000000"
+	res, err := SolidityUnpackString(data, types)
+	if err != nil {
+		log.Println(err)
+	}
+
+	type Recipt struct {
```
Comment (enlight) [inline common/evmcompat/helpers_test.go]: Receipt

Comment (mimiiiz) [reply common/evmcompat/helpers_test.go]: 😢

File: common/evmcompat/helpers.go [hunk 2]
```diff
@@ -191,3 +191,85 @@ func SolidityPackedBytes(pairs []*Pair) ([]byte, error) {
...
+	switch typeString {
+	case "uint8":
+		theInt, err := strconv.ParseInt(partialData, 10, 8)
+		if err != nil {
+			return nil, false, err
+		}
+		return uint8(theInt), false, nil
+
+	case "uint32":
+		theInt, err := strconv.ParseInt(partialData, 10, 32)
```
Comment (enlight) [inline common/evmcompat/helpers.go]: Shouldn't this be `strconv.ParseUint`?

File: common/evmcompat/helpers.go [hunk 3]
```diff
@@ -191,3 +191,85 @@ func SolidityPackedBytes(pairs []*Pair) ([]byte, error) {
...
+		}
+		resp[i] = convertedData
+	}
+	return resp, nil
+}
+
+func TypeConverter(partialData, typeString, dataLeft string, chunkLeft, stringCount int) (interface{}, bool, error) {
+	switch typeString {
+	case "uint8":
+		theInt, err := strconv.ParseInt(partialData, 10, 8)
```
Comment (enlight) [inline common/evmcompat/helpers.go]: Shouldn't this be `strconv.ParseUint`?

File: common/evmcompat/helpers.go [hunk 4]
```diff
@@ -191,3 +191,85 @@ func SolidityPackedBytes(pairs []*Pair) ([]byte, error) {
...
+		}
+		if isString {
+			stringCount += 1
+		}
+		resp[i] = convertedData
+	}
+	return resp, nil
+}
+
+func TypeConverter(partialData, typeString, dataLeft string, chunkLeft, stringCount int) (interface{}, bool, error) {
```
Comment (enlight) [inline common/evmcompat/helpers.go]: `TypeConverter` -> `parseNextValueFromSolidityHexStr`
This function doesn't need to be exported.

File: common/evmcompat/helpers.go [hunk 5]
```diff
@@ -191,3 +191,85 @@ func SolidityPackedBytes(pairs []*Pair) ([]byte, error) {
 
 	return b.Bytes(), nil
 }
+
+func SolidityUnpackString(data string, types []string) ([]interface{}, error) {
```
Comment (enlight) [inline common/evmcompat/helpers.go]: Add comment describing what this function does, and what the `types` are supported.

File: common/evmcompat/helpers.go [hunk 6]
```diff
@@ -191,3 +191,88 @@ func SolidityPackedBytes(pairs []*Pair) ([]byte, error) {
...
+		//substring equal to len
+		i := new(big.Int)
+		stringLen, ok := i.SetString(lenChunk, 16)
+		if !ok {
+			return nil, false, errors.New(fmt.Sprintf("Error parsing big.Int from %s", partialData))
+		}
+		byteString = byteString[:stringLen.Int64()]
+		stringConverted := string(byteString)
+
+		return strings.ToLower(stringConverted), true, nil
```
Comment (enlight) [inline common/evmcompat/helpers.go]: Instead of returning true here return `stringCount+1`, then you don't have to do it in `SolidityUnpackString`

**Pull request merged**